### PR TITLE
DAOS-14020 pool: dup op detection logic disabled, retry tests

### PR DESCRIPTION
**Pull request by kccain**
This is one in a series of patches that will implement
duplicate (retry) detection for pool/container service
(metadata) operations.

With this change, the pool service code is updated to work with
the  previously-added svc_ops KVS added at the root level in a 
previous (container-focused) patch.

Pool service operation handlers are updated to check for duplicate
operations in svc_ops, and (later, when the logic is fully enabled)
save the results of handler execution into svc_ops entries. For now
in this patch, the "save" logic continues to be disabled, just as it
is in the container engine handlers. In duplicate op detection
scenarios, handlers re-read the necessary data to supply it in the
RPC replies.

Container service operation handlers are similarly updated.

The daos_test pool_op_retry() and co_op_retry() tests are updated
to inject faults DAOS_MD_OP_PASS_NOREPLY and DAOS_MD_OP_FAIL_NOREPLY.
And the pool and container service engine handler functions are
updated to react to these fault injections. DAOS_MD_OP_PASS_NOREPLY
causes the handlers to run as usual, and rather than returning rc=0,
force the client to retry the operation. DAOS_MD_OP_FAIL_NOREPLY
causes the handler to fail (e.g., with a simulated -DER_MISC),
but forces the client to retry the operation before it would see the
actual error in the RPC reply.

Also the pool_op_retry() and co_op_retry() functions exercise many
more metadata RPC operations, including some server to server ones
that are triggered by the dmg command. For this, some new dmg
helper functions are added that were previously unavailable
(dmg_pool_evict, dmg_pool_update/delete_ace).

pool and container client code is restructured a bit related to
task private pointer assignment and clearing. Also the
pool/cont_req_prepare() functions had a bug in the location
of the assignment of args->cra_tpriv that is fixed in this change.

The NLT test maximum log size parameter is increased, since the
additional vos operations used for rdb_tx_lookup, etc. in svc_ops
cause more logging to occur and otherwise push the log sizes over
the previous limit.

Features: pool
Allow-unstable-test: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/container/cli.c [hunk 1]
```diff
@@ -2415,8 +2416,11 @@ cont_req_cleanup(enum creq_cleanup_stage stage, bool free_tpriv, struct cont_req
 	case CLEANUP_RPC:
 		crt_req_decref(args->cra_rpc);
 	case CLEANUP_TASK_PRIV:
-		if (free_tpriv)
+		if (free_tpriv) {
 			D_FREE(args->cra_tpriv);
+			args->cra_tpriv = NULL;
```
Comment (wangshilong) [inline src/container/cli.c:2421]: D_FREE() will set NULL, you don't have to do it.

File: src/pool/cli.c [hunk 2]
```diff
@@ -2557,8 +2558,11 @@ pool_req_cleanup(enum preq_cleanup_stage stage, bool free_tpriv, struct pool_req
 	case CLEANUP_RPC:
 		crt_req_decref(args->pra_rpc);
 	case CLEANUP_TASK_PRIV:
-		if (free_tpriv)
+		if (free_tpriv) {
 			D_FREE(args->pra_tpriv);
+			args->pra_tpriv = NULL;
```
Comment (wangshilong) [inline src/pool/cli.c:2563]: args->pra_tpriv = NULL is unnecessary.

File: src/pool/rpc.h [hunk 3]
```diff
@@ -604,6 +253,33 @@ pool_connect_in_set_data(crt_rpc_t *rpc, uint64_t pci_flags, uint64_t pci_query_
...
+	    (uint64_t)(pqi_query_bits)CRT_VAR)
+
+#define DAOS_ISEQ_POOL_QUERY_V6 /* input fields */                                                 \
+	((struct pool_op_v6_in)(pqi_op)CRT_VAR)((crt_bulk_t)(pqi_map_bulk)CRT_VAR)(                \
+	    (uint64_t)(pqi_query_bits)CRT_VAR)
+
+#define DAOS_OSEQ_POOL_QUERY /* output fields */                                                   \
+	((struct pool_op_out)(pqo_op)CRT_VAR)((daos_prop_t)(pqo_prop)CRT_PTR)(                     \
+	    (struct daos_pool_space)(pqo_space)CRT_RAW)(                                           \
+	    (struct daos_rebuild_status)(pqo_rebuild_st)CRT_RAW) /* only set on -DER_TRUNC */      \
```
Comment (wangshilong) [inline src/pool/rpc.h:276]: These macro looks hard to read, why not keep space between var?

File: src/pool/srv_pool.c [hunk 4]
```diff
@@ -2658,6 +2668,184 @@ pool_prop_read(struct rdb_tx *tx, const struct pool_svc *svc, uint64_t bits,
...
+	case POOL_FILTER_CONT:
+	case POOL_PROP_GET:
+	case POOL_RANKS_GET:
+	/* opcodes not handled by pool service */
+	case POOL_TGT_QUERY_MAP:
+	case POOL_TGT_DISCONNECT:
+	case POOL_TGT_QUERY:
+	case POOL_ADD_TGT:
+	case POOL_TGT_DISCARD:
+		is_write = false;
```
Comment (wangshilong) [inline src/pool/srv_pool.c:2693]: [defect] missed a break?

File: src/container/srv_container.c [hunk 5]
```diff
@@ -5273,6 +5311,12 @@ cont_op_save_dup(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl, struct cont_sv
 	crt_opcode_t              opc = opc_get(rpc->cr_opc);
 	int                       rc  = 0;
 
+	/* Test here for simpler caller invocation code */
+	if (dup_op)
+		goto out;
```
Comment (liw) [inline src/container/srv_container.c:5316]: [Nit; not requesting changes] This reminds me that the `cont_op_save_dup` name sounds like "save duplicate" and this "if" statement seems to contradict the name, whereas what we really want is
```
if not a duplicate
    save result
```
I hope I didn't misunderstand this function.

File: src/pool/srv_pool.c [hunk 6]
```diff
@@ -3135,26 +3333,40 @@ ds_pool_connect_handler(crt_rpc_t *rpc, int handler_version)
...
-	out->pco_op.po_map_version = ds_pool_get_version(svc->ps_pool);
+
 out_lock:
 	ABT_rwlock_unlock(svc->ps_lock);
 	rdb_tx_end(&tx);
 	if (rc == 0 && transfer_map) {
 		rc = ds_pool_transfer_map_buf(map_buf, map_version, rpc, bulk,
 					      &out->pco_map_buf_size);
-		/** TODO: roll back tx if transfer fails? */
+		/** TODO: roll back tx if transfer fails? Perhaps rdb_tx_discard()? */
```
Comment (liw) [inline src/pool/srv_pool.c:3366]: As you know, `rdb_tx_discard` is meant for discarding writes in an uncommitted RDB TX; out-of-date comment, perhaps?

File: src/container/srv_container.c [hunk 7]
```diff
@@ -5441,8 +5494,20 @@ cont_op_with_svc(struct ds_pool_hdl *pool_hdl, struct cont_svc *svc,
 			ds_cont_prop_iv_update(svc, in->ci_uuid);
 	}
 
+	if ((rc == 0) && !dup_op && fi_pass_noreply) {
+		rc = -DER_TIMEDOUT;
+		D_DEBUG(DB_MD, DF_UUID ": fault injected: DAOS_MD_OP_PASS_NOREPLY\n",
+			DP_UUID(in->ci_uuid));
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/container/srv_container.c:5500]: maybe we can do all of these (across both this file and srv_container.c, including other fault injections e.g., the CORPC related ones in a subsequent patch?

File: src/container/cli.c (hunk 1 above)
Comment (kccain) [reply src/container/cli.c:2421]: will fix

File: src/pool/rpc.h (hunk 3 above)
Comment (kccain) [reply src/pool/rpc.h:276]: comes from clang-format. There may be a recipe available for clang-format to (temporarily) not flag these macros with the human-readable spacing that we need. I will fix them, but perhaps in a separate PR if it is too complicated to apply the technique along with the other changes I'm making in this patch.

File: src/pool/srv_pool.c (hunk 4 above)
Comment (kccain) [reply src/pool/srv_pool.c:2693]: will fix, thanks!

File: src/container/srv_container.c [hunk 8]
```diff
@@ -5359,72 +5409,75 @@ cont_op_with_svc(struct ds_pool_hdl *pool_hdl, struct cont_svc *svc,
...
 	rc = get_metadata_times(&tx, cont, false /* otime */, update_mtime, NULL /* times */);
+	if (rc != 0)
+		goto out_commit;
+
+out_commit:
+	if ((rc == 0) && !dup_op && fi_fail_noreply)
+		rc = cont_op_save_dup(&tx, pool_hdl, svc, rpc, dup_op, cont_proto_ver, -DER_MISC,
+				      &op_val);
+	else
+		rc = cont_op_save_dup(&tx, pool_hdl, svc, rpc, dup_op, cont_proto_ver, rc, &op_val);
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/container/srv_container.c:5469]: I think this and the other comment are related https://github.com/daos-stack/daos/pull/13396#discussion_r1411744642

I might like to do something like this
```
if (rc == 0 && !dup_op && fi_fail_noreply)
    rc = -DER_MISC;
if (!dup_op)
    rc = cont_op_save_dup(&tx, pool_hdl, svc, rpc, cont_proto_ver, rc, &op_val);

// rc is understood to be 0 if dup_op is true
```

**Review by liw — APPROVED** [review]

File: src/container/srv_container.c (hunk 5 above)
Comment (liw) [reply src/container/srv_container.c]: The revised code looks much easier to read, I think. Thanks.

File: src/container/srv_container.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (liw) [reply src/container/srv_container.c:5494]: Sure.

File: src/pool/rpc.h (hunk 3 above)
Comment (kccain) [reply src/pool/rpc.h:276]: In a separate draft PR based on this branch, I've tried to manually fix up the macros affected by clang-format. The code there manually uses tabs to make the macros readable, and uses clang-format off / on comments to mark what sections of the header file should be ignored (or processed) by clang-format.

https://github.com/daos-stack/daos/pull/13440/

**Review by wangshilong — APPROVED** [review]

**Comment by kccain** [comment]
Test build 6 just took days to complete successfully. And now there is a conflict with src/include/daos/common.h

So I have to resolve the conflict and push the result for more testing...

**Comment by kccain** [comment]
Build 6 passed all per-PR tests.
Earlier iterations of the patch were tested with Features: pool successfully

Patch just updated to  solve a pretty straightforward file conflict in src/include/common/daos.h and build 7 is in progress https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13396/7/pipeline-graph/

@daos-stack/daos-gatekeeper I propose proactively reviewing this patch, and monitoring the build 7 progress to a reasonable point (say, through "Test" stage i.e.., functional VM testing?) and land if all looks reasonable in the review.

**Review by wangshilong — APPROVED** [review]

**Comment by daltonbohning** [comment]
> VerifyPoolSpace appears to fail on master in the same way. Last week I prepared a separate test-only PR 13414 to explore some of these failures, whether they would happen on master independent of this patch. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13414/6/testReport/junit/FTEST_pool/VerifyPoolSpace/Test_Hardware___Functional_Hardware_Medium_MD_on_SSD___1___pool_verify_space_py_VerifyPoolSpace_test_verify_pool_space_run_container_hosts_ior_pool_rank_0_pool_rank_0_1_2_pool_rank_1_2_pool_rank_1_a_pool_rank_1_b_pool_rank_1_c_server_config_control_metadata_engines_0_storage_0_1_2_a886/
> 
> Maybe it needs a ticket, not sure. There had been reports of this test failing out on the feature branch in the past https://daosio.atlassian.net/browse/DAOS-14546

Ah, it's in the MD on SSD stage, and a weekly test, which doesn't have recurring coverage yet. We can ignore those for now

**Review by liw — APPROVED** [review]

**Comment by kccain** [comment]
Jenkins build 7 is successfully through the Test stage and has been waiting to run on hardware clusters for almost a day. I think it's overkill to wait for that, though. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13396/7/

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium Verbs Provider completed with status UNSTABLE. https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13396/7/testReport/

this appears to be an instance of existing failure https://daosio.atlassian.net/browse/DAOS-14353

**Pull request merged**